src/hydranfc_v2/hydranfc_v2_bbio_card_emulator.c

```diff
@@ -0,0 +1,253 @@
+#include "rfal_analogConfig.h"
+#include "rfal_rf.h"
+#include "rfal_nfca.h"
+#include "rfal_nfcb.h"
+#include "rfal_nfcv.h"
+
+#include <string.h>
+
+static volatile int irq_count;
+static volatile int irq;
+static volatile int irq_end_rx;
+
+static void (* st25r3916_irq_fn)(void) = NULL;
+
+extern sUserTagProperties user_tag_properties;
+extern void hydranfc_ce_set_processCmd_ptr(void * ptr);
+
+/* Triggered when the Ext IRQ is pressed or released. */
+static void extcb1(void * arg)
+{
+	(void) arg;
+
+	if(st25r3916_irq_fn != NULL)
+		st25r3916_irq_fn();
+
+	irq_count++;
+	irq = 1;
+}
+
+extern t_mode_config mode_con1;
+static t_hydra_console * g_con;
+
+static ReturnCode hydranfc_v2_init_RFAL(t_hydra_console *con)
+{
+	ReturnCode err;
+	/* RFAL initalisation */
+	rfalAnalogConfigInitialize();
+	err = rfalInitialize();
+	if(err != ERR_NONE) {
+		cprintf(con, "hydranfc_v2_init_RFAL rfalInitialize() error=%d\r\n", err);
+		return err;
+	}
+
+	/* DPO setup */
+#ifdef DPO_ENABLE
+	rfalDpoInitialize();
+	rfalDpoSetMeasureCallback( rfalChipMeasureAmplitude );
+	err = rfalDpoTableWrite(dpoSetup,sizeof(dpoSetup)/sizeof(rfalDpoEntry));
+	if(err != ERR_NONE) {
+		cprintf(con, "hydranfc_v2_init_RFAL rfalDpoTableWrite() error=%d\r\n", err);
+		return err;
+	}
+	rfalDpoSetEnabled(true);
+	rfalSetPreTxRxCallback(&rfalPreTransceiveCb);
+#endif
+	return err;
+}
+
+
+static bool init_gpio_spi_nfc(t_hydra_console *con)
+{
+	/*
+	 * Initializes the SPI driver 2. The SPI2 signals are routed as follow:
+	 * ST25R3916 IO4_CS SPI mode / HydraBus PC1 - NSS
+	 * ST25R3916 DATA_CLK SPI mode / HydraBus PB10 - SCK
+	 * ST25R3916 IO6_MISO SPI mode / HydraBus PC2 - MISO
+	 * ST25R3916 IO7_MOSI SPI mode / HydraBus PC3 - MOSI
+	 * Used for communication with ST25R3916 in SPI mode with NSS.
+	 */
+	mode_con1.proto.config.spi.dev_gpio_pull = MODE_CONFIG_DEV_GPIO_NOPULL;
+	//mode_con1.proto.config.spi.dev_speed = 5; /* 5 250 000 Hz */
+	mode_con1.proto.config.spi.dev_speed = 6; /* 10 500 000 Hz */
+	mode_con1.proto.config.spi.dev_phase = 1;
+	mode_con1.proto.config.spi.dev_polarity = 0;
+	mode_con1.proto.config.spi.dev_bit_lsb_msb = DEV_FIRSTBIT_MSB;
+	mode_con1.proto.config.spi.dev_mode = DEV_MASTER;
+	bsp_spi_init(BSP_DEV_SPI2, &mode_con1.proto);
+
+	/*
+	 * Initializes the SPI driver 1. The SPI1 signals are routed as follows:
+	 * Shall be configured as SPI Slave for ST25R3916 NFC data sampling on MOD pin.
+	 * NSS. (Not used use Software).
+	 * ST25R3916 MCU_CLK pin28 output / HydraBus PA5 - SCK.(AF5) => SPI Slave CLK input (Sniffer mode/RX Transparent Mode)
+	 * ST25R3916 ST25R3916 MOSI SPI pin31 (IN) / HydraBus PA6 - MISO.(AF5) (MCU TX Transparent Mode)
+	 * ST25R3916 ST25R3916 MISO_SDA pin32 (OUT) / HydraBus PA7 - MOSI.(AF5) => SPI Slave MOSI input (Sniffer mode/RX Transparent Mode)
+	 */
+	/* spiStart() is done in sniffer see sniffer.c */
+	/* HydraBus SPI1 Slave CLK input */
+	palSetPadMode(GPIOA, 5, PAL_MODE_ALTERNATE(5) | PAL_STM32_OSPEED_MID1);
```

Comment:
To be refactored using latest code from master https://github.com/hydrabus/hydrafw_hydranfc_shield_v2/blob/master/src/hydranfc_v2/hydranfc_v2_nfc_mode.c#L163
The idea is to remove the maximum of ChibiOS stuff related to GPIO (except Interrupt to be done later) and use bsp_gpio_init()